service/orchestrator/assessment_results.go

```diff
@@ -34,41 +34,53 @@ import (
 	"google.golang.org/grpc/status"
 
 	"clouditor.io/clouditor/api/assessment"
+	"clouditor.io/clouditor/internal/util"
 	"clouditor.io/clouditor/service"
 )
 
 // ListAssessmentResults is a method implementation of the orchestrator interface
 func (svc *Service) ListAssessmentResults(ctx context.Context, req *assessment.ListAssessmentResultsRequest) (res *assessment.ListAssessmentResultsResponse, err error) {
 	var values = maps.Values(svc.results)
-	var filtered []*assessment.AssessmentResult
+	var filtered_values []*assessment.AssessmentResult
 	var allowed []string
 	var all bool
 
 	// Check, if the current filter is valid according to the authorization strategy. Omitting the cloud service ID is
 	// only allowed, if one can access *all* the cloud services.
 	all, allowed = svc.authz.AllowedCloudServices(ctx)
-	if !all && req.FilteredCloudServiceId == "" {
+	if !all && util.Deref(req.FilteredCloudServiceId) == "" {
 		return nil, service.ErrPermissionDenied
 	}
 
 	// Furthermore, the content of the filtered cloud service ID must be in the list of allowed cloud service IDs,
 	// unless one can access *all* the cloud services.
-	if !all && req.FilteredCloudServiceId != "" && !slices.Contains(allowed, req.FilteredCloudServiceId) {
+	if !all && util.Deref(req.FilteredCloudServiceId) != "" && !slices.Contains(allowed, util.Deref(req.FilteredCloudServiceId)) {
 		return nil, service.ErrPermissionDenied
 	}
 
 	for _, v := range values {
-		if req.FilteredCloudServiceId != "" && v.CloudServiceId != req.FilteredCloudServiceId {
+		// Check for filtered cloud service ID
+		if util.Deref(req.FilteredCloudServiceId) != "" && v.CloudServiceId != util.Deref(req.FilteredCloudServiceId) {
 			continue
 		}
 
-		filtered = append(filtered, v)
+		// Check for filter compliant
+		if req.FilteredCompliant != nil && util.Deref(req.FilteredCompliant) != v.Compliant {
+			continue
+		}
+
+		// Check for filter metric ID
+		if req.FilteredMetricId != nil && util.Deref(req.FilteredMetricId) != v.MetricId {
```

Comment:
The nil check is also done in `util.Deref`. Maybe it is redundant here?
I guess you want to prevent a check if the optional value is not set anyway but then we don't need the `Deref` here if you check it right before.

Reply:
See answer above.